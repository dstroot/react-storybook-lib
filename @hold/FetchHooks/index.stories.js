import React from "react";
import { storiesOf } from "@storybook/react";

// components
import FetchHooks from "../FetchHooks";

// FetchHooks section
storiesOf("FetchHooks", module)
  .addParameters({
    // default for all stories in this book
    info: {
      inline: true,
      header: false,
      text: `
        Fetch data using hooks.
      `
    }
  })
  .add("show FetchHooks", () => <FetchHooks />);
