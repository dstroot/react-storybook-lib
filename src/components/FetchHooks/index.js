import React, { useState, useEffect } from 'react';
import axios from 'axios';

const useDataApi = (initialUrl, initialData) => {
  // state management
  const [data, setData] = useState(initialData);
  const [url, setUrl] = useState(initialUrl);
  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);

  // make API call
  const fetchData = async () => {
    setIsError(false);
    setIsLoading(true);

    try {
      const result = await axios(url);
      setData(result.data);
    } catch (error) {
      setIsError(true);
      console.log(error.message);
    }

    setIsLoading(false);
  };

  // useEffect hook to call API
  useEffect(
    () => {
      fetchData();
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [url] // only run if url changes
  );

  const doGet = (event, url) => {
    event.preventDefault();
    setUrl(url);
  };

  return { data, isLoading, isError, doGet };
};

function FetchHooks() {
  const [query, setQuery] = useState('redux');
  const { data, isLoading, isError, doGet } = useDataApi(
    'http://hn.algolia.com/api/v1/search?query=redux',
    { hits: [] }
  );

  return (
    <div className="container">
      <form
        className="form-inline mt-5"
        onSubmit={event =>
          doGet(event, `http://hn.algolia.com/api/v1/search?query=${query}`)
        }
      >
        <label className="sr-only" for="query">
          Query
        </label>
        <input
          type="text"
          class="form-control mb-2 mr-sm-2"
          id="query"
          placeholder={query}
          onChange={event => setQuery(event.target.value)}
        />
        <button className="btn btn-primary mb-2" type="submit">
          Search
        </button>
      </form>

      {isError && <div className="text-danger">Something went wrong ...</div>}

      {isLoading ? (
        <div className="text-muted">Loading ...</div>
      ) : (
        <ul>
          {data.hits.map(item => (
            <li key={item.objectID}>
              <a href={item.url}>{item.title}</a>
            </li>
          ))}
        </ul>
      )}
    </div>
  );
}

export default FetchHooks;
