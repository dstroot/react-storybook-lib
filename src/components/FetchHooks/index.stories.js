import React from 'react';
import { storiesOf } from '@storybook/react';

// components
import FetchHooks from '../FetchHooks';

// FetchHooks section
storiesOf('FetchHooks', module).add('show FetchHooks', () => <FetchHooks />);
