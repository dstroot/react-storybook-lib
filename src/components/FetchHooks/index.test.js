import React from 'react';
import renderer from 'react-test-renderer';
import FetchHooks from '../FetchHooks';

describe('FetchHooks', () => {
  it('it should render', () => {
    const component = renderer.create(<FetchHooks />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
