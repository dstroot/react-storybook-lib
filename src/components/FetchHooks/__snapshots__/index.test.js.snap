// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchHooks it should render 1`] = `
<div
  className="container"
>
  <form
    className="form-inline mt-5"
    onSubmit={[Function]}
  >
    <label
      className="sr-only"
      for="query"
    >
      Query
    </label>
    <input
      class="form-control mb-2 mr-sm-2"
      id="query"
      onChange={[Function]}
      placeholder="redux"
      type="text"
    />
    <button
      className="btn btn-primary mb-2"
      type="submit"
    >
      Search
    </button>
  </form>
  <ul />
</div>
`;
